go/shuffle/shuffle.go

```diff
@@ -0,0 +1,150 @@
+package shuffle
+
+import (
+	"encoding/json"
+	"reflect"
+
+	pf "github.com/estuary/flow/go/protocols/flow"
+	pr "github.com/estuary/flow/go/protocols/runtime"
+)
+
+// shuffle is an internal description of a source shuffle.
+// It's not intended to be used outside of the `shuffle` package,
+// or to be persisted / serialized, and its layout and semantics
+// may change.
+type shuffle struct {
+	// Should documents be filtered on their r-clocks?
+	// This is only true for read-only derivation transforms.
+	filterRClocks bool
+	// Path metadata suffix attached to journals read by this shuffle.
+	journalReadSuffix string
+	// Priority of this shuffle with respect to others of the derivation.
+	priority uint32
+	// Read delay of this shuffle with respect to others of the derivation.
+	readDelaySeconds uint32
+	// Key of this shuffle. If empty, then `usesLambda` is true.
+	shuffleKey []string
+	// Partitioned projection fields which fully cover the shuffle key.
+	shuffleKeyPartitionFields []string
+	// Name of the sourced collection.
+	sourceCollection pf.Collection
+	// Partition selector of the sourced collection.
+	sourcePartitions pf.LabelSelector
+	// JSON pointer of source document UUIDs.
+	sourceUuidPtr string
+	// Shuffle key is dynamically computed from derivation transform source documents.
+	usesLambda bool
+	// Shuffle key is the same as the source collection key.
+	usesSourceKey bool
+	// If non-nil, validate the schema on read.
+	validateSchema json.RawMessage
+}
+
+func derivationShuffles(task *pf.CollectionSpec) []shuffle {
+	if task.Derivation == nil {
+		return nil
```

Comment:
Is there any reason we shouldn't panic here (and in `materializationShuffles`)? Seems like this should never be nil IRL. Perhaps just removing the check could work too.